### PR TITLE
[Tokenomics] SettleSessionAccounting first implementation (emissions=burn)

### PR DESCRIPTION
**Pull request by Olshansk**
## Summary

Implement `SettleSessionAccounting` where emissions=burn.

## Issue

- #243 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Run E2E tests locally**: `make test_e2e`
- [ ] **Run E2E tests on DevNet**: Add the `devnet-test-e2e` label to the PR. This is VERY expensive, only do it after all the reviews are complete.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

**Comment by Olshansk** [comment]
@bryanchriswhite @h5law I left a couple `TODO_IN_THIS_PR`, but it's ready to get a first round of review in parallel.

File: x/tokenomics/keeper/settle_session_accounting.go [hunk 1]
```diff
@@ -2,15 +2,138 @@ package keeper
...
+	params := k.GetParams(ctx)
+
+	// Calculate the amount of tokens to mint & burn
+	upokt := sdk.NewInt(int64(claimComputeUnits * params.ComputeUnitsToTokensMultiplier))
+	upoktCoin := sdk.NewCoin("upokt", upokt)
+	upoktCoins := sdk.NewCoins(upoktCoin)
+
+	logger.Info(fmt.Sprintf("%d compute units equate to %d uPOKT for session %s", claimComputeUnits, upokt, sessionHeader.SessionId))
+
+	// NB: We are doing a mint & burn + transfer, instead of a simple transfer
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting.go:80]: Nice comment

File: x/tokenomics/keeper/settle_session_accounting.go [hunk 2]
```diff
@@ -2,15 +2,138 @@ package keeper
...
+	); err != nil {
+		return types.ErrTokenomicsApplicationModuleFeeFailed
+	}
+
+	logger.Info(fmt.Sprintf("sent %d uPOKT to supplier with address %s", upokt, supplierAddress))
+
+	// Verify that the application has enough uPOKT to pay for the services it consumed
+	if application.Stake.IsLT(upoktCoin) {
+		logger.Error(fmt.Sprintf("THIS SHOULD NOT HAPPEN. Application with address %s needs to be charged more than it has staked: %v > %v", applicationAddress, upoktCoins, application.Stake))
+		// TODO_BLOCKER(@Olshansk, @RawthiL): The application was over-serviced in the last session so it basically
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting.go:104]: Just a thought but rate limiting at the supplier level to ensure the application can pay for their services moving forward could be a play - nothing actionable something to think on - @RawthiL had mentioned something like this before

File: x/tokenomics/keeper/settle_session_accounting.go [hunk 3]
```diff
@@ -2,15 +2,138 @@ package keeper
...
+	if err := k.bankKeeper.MintCoins(ctx, suppliertypes.ModuleName, upoktCoins); err != nil {
+		return types.ErrTokenomicsApplicationModuleFeeFailed
+	}
+
+	logger.Info(fmt.Sprintf("minted %d uPOKT in the supplier module", upokt))
+
+	// Sent the minted coins to the supplier
+	if err := k.bankKeeper.SendCoinsFromModuleToAccount(
+		ctx, suppliertypes.ModuleName, supplierAddress, upoktCoins,
+	); err != nil {
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting.go:95]: Just a thought - should we attempt to rollback any minting done before returning an error?

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+			desc:        "Nil Root",
+			root:        nil,
+			errExpected: true,
+		},
+		{
+			desc:        "Less than 40 bytes",
+			root:        []byte("less than 40 bytes"),
+			errExpected: true,
+		},
+		// TODO_IN_THIS_PR_DISCUSS: Should this error? Should it be on the SMT side?
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting_test.go:84]: This should be checked here IMO - the SMT **cannot** produce a root that doesn't coincide with the hasher used and whether its a SMT/SMST. Meaning that if it isn't 40 bytes the node is erroring.

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+		errExpected bool
+	}{
+		{
+			desc:        "Nil Root",
+			root:        nil,
+			errExpected: true,
+		},
+		{
+			desc:        "Less than 40 bytes",
+			root:        []byte("less than 40 bytes"),
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting_test.go:81]:
```suggestion
			root:        make([]byte, 39), // Less than 40 bytes
```

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+			errExpected: true,
+		},
+		// TODO_IN_THIS_PR_DISCUSS: Should this error? Should it be on the SMT side?
+		// {
+		// 	desc:      "More than 40 bytes",
+		// 	root:      []byte("more than 40 bytes, meaning this is way too long"),
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting_test.go:87]:
```suggestion
		// 	root:      make([]byte, 41), // More than 40 bytes
```

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+		// 	errExpected: true,
+		// },
+		// TODO_IN_THIS_PR_DISCUSS: Should this error? Should it be on the SMT side?
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting_test.go:90]: I'm not sure what this means, `make([]byte, 40)` aka a 40-byte slice with all 0s represents an empty tree with a 0 sum. This is valid iff. the node did no work in the session.

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+		// {
+		// 	desc: "40 bytes but empty",
+		// 	root: func() []byte {
+		// 		root := [40]byte{}
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting_test.go:94]:
```suggestion
		// 		root := make([]byte, 40) // 40-byte slice of all 0s
```

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+			// Use defer-recover to catch any panic
+			defer func() {
+				if r := recover(); r != nil {
+					t.Errorf("Test panicked: %s", r)
+				}
+			}()
+
+			// Setup claim by copying the baseClaim and updating the root
+			claim := baseClaim(appAddr, supplierAddr, 0)
+			claim.RootHash = smt.MerkleRoot(tc.root[:])
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting_test.go:130]: Should this not be:
```suggestion
			claim.RootHash = smt.MerkleRoot(tc.root())
```

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+		},
+		RootHash: smstRootWithSum(42),
+	}
+}
+
+func smstRootWithSum(sum uint64) smt.MerkleRoot {
+	var root [40]byte
+	copy(root[:32], []byte("This is exactly 32 characters!!!"))
+	binary.BigEndian.PutUint64(root[32:], sum)
+	return smt.MerkleRoot(root[:])
```
Comment (h5law) [inline x/tokenomics/keeper/settle_session_accounting_test.go:272]:
```suggestion
	root := make([]byte, 40)
	copy(root[:32], []byte("This is exactly 32 characters!!!"))
	binary.BigEndian.PutUint64(root[32:], sum)
	return smt.MerkleRoot(root)
```

**Review by h5law — CHANGES_REQUESTED** [review]
Nice left some comments on the tests - but things are shaping up nicely! Make sure you fix `x/tokenomics/genesis_test.go` as `keepertest.TokenomicsKeeper(t)` now returns 4 values.

File: x/session/types/session_header.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+	sdkerrors "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// TODO_TECHDEBT: Make sure this is used everywhere we validate components
+// of the session header.
+func (sh *SessionHeader) ValidateBasic() error {
+	// Validate the application address
+	if _, err := sdk.AccAddressFromBech32(sh.ApplicationAddress); err != nil {
+		return sdkerrors.Wrapf(ErrSessionInvalidAppAddress, "invalid application address: %s; (%v)", sh.ApplicationAddress, err)
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/session/types/session_header.go:13]: Resolving this, but please also see the comment I left:

```
// TODO_TECHDEBT: Do a source code wise find-replace using regex pattern match
// of `sdkerrors\.Wrapf\(([a-zA-Z]+), ` with `$1.Wrapf(`
```

File: testutil/keeper/tokenomics.go [hunk 12]
```diff
@@ -16,27 +16,76 @@ import (
...
+	// Initialize the codec and other necessary components
 	registry := codectypes.NewInterfaceRegistry()
 	cdc := codec.NewProtoCodec(registry)
+	ctrl := gomock.NewController(t)
 
+	// The on-chain governance address
 	authority := authtypes.NewModuleAddress("gov").String()
 
-	ctrl := gomock.NewController(t)
+	// Prepare the test application
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply testutil/keeper/tokenomics.go:55]: _tl;dr Thought about it but do not believe it is worth the tradeoff in this chase._

cc @bryanchriswhite for additional thoughts or pushback.

---

I can steelman both sides here. 

Firstly, I always bias to writing less code wherever possible: less code => less bugs => less to maintain => less to learn => more time on other things.

With that said,  Complex state management in mocks is nice when you're testing some business logic, but in this specific instance, I feel that:
1. I would spend a lot of time implementing this business logic, which will be quickly deprecated by in-memory testing network
2. A single e2e test might provide just as much benefit
3. I documented the unit tests (in x/tokenomics/keeper/settle_session_accounting_test.go) that I want to add once we have (1)
4. There is the risk of that logic having bugs too.

It's a very delicate balance and don't think there's a definitive answer to when to do which one.

File: x/tokenomics/keeper/settle_session_accounting.go (hunk 3 above)
Comment (Olshansk) [reply x/tokenomics/keeper/settle_session_accounting.go:95]: This did cross my mind and we definitely need to do if it's not atomic.

Adding the following comment in the meantime to block merging

```
// TODO_IN_THIS_PR: Rollback state changes if we find out that this function is not
// atomic.if this function is not atomic.
```

---

CCing @tac0turtle - Do you know at what scope / granularity keeper functions are atomic? I.e. if an error is returned, do the state changes get rolled back?

File: x/tokenomics/keeper/settle_session_accounting_test.go (hunk 9 above)
Comment (Olshansk) [reply x/tokenomics/keeper/settle_session_accounting_test.go:130]: No. `root        []byte // smst.MerkleRoot`

I call the function during the test prep so it's `root` not `rootSetup`. I think its clearer.

File: x/tokenomics/keeper/settle_session_accounting.go (hunk 3 above)
Comment (Olshansk) [reply x/tokenomics/keeper/settle_session_accounting.go:113]: Thanks Marko!

Screenshot 2024-01-26 at 9 16 22 AM

**Review by red-0ne — APPROVED** [review]
Left some NIT comments but approving.

Great to see we have this working finally!

File: testutil/keeper/tokenomics.go [hunk 13]
```diff
@@ -16,27 +16,71 @@ import (
...
+		GetApplication(gomock.Any(), gomock.Eq(application.Address)).
+		Return(application, true).
+		AnyTimes()
+	mockApplicationKeeper.EXPECT().
+		GetApplication(gomock.Any(), gomock.Not(application.Address)).
+		Return(apptypes.Application{}, false).
+		AnyTimes()
+	mockApplicationKeeper.EXPECT().
+		SetApplication(gomock.Any(), gomock.Any()).
+		AnyTimes()
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply testutil/keeper/tokenomics.go:74]: This is a good idea, but we reuse the same mocks in different tests so I'm adding a TODO and leaving as is.

**Pull request merged**